app/models/user.rb

```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  before_save{email.downcase!}
+  validates :name, presence: true, length: {maximum: Settings.maximum}
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
+  validates :email, presence: true, length: {maximum: Settings.maximum2},
+    format: {with: VALID_EMAIL_REGEX},
+    uniqueness: {case_sensitive: false}
+  has_secure_password
```

Comment:
lệnh trên có tác dụng gì?

Reply:
validates là để xác nhận :name có hiện diện hay k với agurment presence: true ạ, và tối đa 50 kí tự 
valid_email_regex là biểu thức chính qui. 
validates ở duới thì cũng xác nhận :email có hiện diện hay không với argument  presence: true. tối đa 255 kí tự
với định dạng phù hợp là biểu thức chính qui ở trên ạ
has_secure_password là sử dụng bcrypt để băm mật khẩu để bảo mật. Lưu hash của mật khẩu thay vì mật khẩu thực. Khi xác thực người dùng, lấy mã băm của mật khẩu và so sánh nó với hash trên cơ sở dữ liệu.

Reply:
Validates là các thao tác kiểm tra dữ liệu trước khi lưu một object vào database.
không cho phép bỏ trống email và tối đa 255 kí tự với định dạng phù hợp là biểu thức chính qui valid_email_regex ở trên ạ. email phải là duy nhất. không phân biệt hoa thường
has_secure_password là sử dụng bcrypt để băm mật khẩu để bảo mật. Lưu hash của mật khẩu thay vì mật khẩu thực. Khi xác thực người dùng, lấy mã băm của mật khẩu và so sánh nó với hash trên cơ sở dữ liệu.

Reply:
vẫn chưa đủ nhé, đọc kĩ lại xem nào

Reply:
before_save là 1 callback để kiểm tra logic trước khi phuơng thức đc lưu vào ạ
email.downcase kiểm tra xem email có viết thường không trước khi lưu vào database
has_secure_password có các chức năng như 1 method authenticate trả về khi nhập mật khẩu đúng, nếu không trả về thì false, khả năng lưu thuộc tính password_digest băm hợp nhất an toàn vào cơ sở dữ liệu, một cặp thuộc tính (password và password_confirmation), bao gồm xác nhận hiện diện khi tạo đối tượng và xác nhận hợp lệ yêu cầu chúng khớp
Yêu cầu duy nhất cho has_secure_password hoạt động là cho mô hình tương ứng có một thuộc tính được gọi là password_digest.

Reply:
chỗ này a chỉ hỏi về has_secure_password thôi, nhưng phần nói thêm về callback chưa đúng lắm r. chỉ để kiểm tra thôi à? nói lại phần đó xem nào.

Reply:
Callback là các phương thức được gọi tới vào một thời điểm nào đó trong vòng đời của một đối tượng. Callback thường được dùng để thực thi các phương thức logic trước hoặc sau khi đối tượng được created, updated, saved, deleted, validated hoặc load từ cơ sở dữ liệu.